lib/aptible/cli/helpers/s3_log_helpers.rb

```diff
@@ -0,0 +1,218 @@
+            # docker file log rotation. So we eliminate some useless items
+            # and then work from the beginning or end of the remaining to find
+            # known elements, ignoring any .1 .2 (or none at all) extension
+            # found in the middle of the file name. EG:
+            # ['container_id', 'start_time', 'end_time']
+            # or
+            # ['container_id', '.1', 'start_time', 'end_time']]
+            split_by_dot = file_name.split('.') - %w(log gz archived)
+            properties[:container_id] = split_by_dot.first.delete!('-json')
+            properties[:start_time] = utc_datetime(split_by_dot[-2])
+            properties[:end_time] = utc_datetime(split_by_dot[-1])
+          else
+            m = "Cannot determine aptible log naming schema from #{file}"
+            raise Thor::Error, m
+          end
+          properties
+        end
+
+        def decrypt_and_translate_s3_file(file, enc_key, region, bucket, path)
+          # AWS warns us about using the legacy encryption schema
+          s3 = Kernel.silence_warnings do
+            Aws::S3::EncryptionV2::Client.new(
+              encryption_key: enc_key, region: region,
+              key_wrap_schema: :aes_gcm,
+              content_encryption_schema: :aes_gcm_no_padding,
+              security_profile: :v2_and_legacy
+            )
+          end
+
+          # Just write it to a file directly
+          location = File.join(path, file.split('/').drop(4).join('/'))
+          FileUtils.mkdir_p(File.dirname(location))
+          File.open(location, 'wb') do |f|
+            CLI.logger.info location
+            # Is this memory efficient?
+            s3.get_object(bucket: bucket, key: file, response_target: f)
+          end
+        end
+
+        def find_s3_files_by_string_match(region, bucket, stack, strings)
+          # This function just regex matches a provided string anywhwere
+          # in the s3 path
+          begin
+            stack_logs = s3_client(region).bucket(bucket)
+                                          .objects(prefix: stack)
+                                          .map(&:key)
+          rescue => error
+            raise Thor::Error, error.message
+          end
+          strings.each do |s|
+            stack_logs = stack_logs.select { |f| f =~ /#{s}/ }
+          end
+          stack_logs
+        end
+
+        def find_s3_files_by_attrs(region, bucket, stack,
+                                   attrs, time_range = nil)
+          # This function uses the known path schema to return files matching
+          # any provided criteria. EG:
+          # * attrs: { :type => 'app', :id => 123 }
+          # * attrs: { :container_id => 'deadbeef' }
+
+          begin
+            stack_logs = s3_client(region).bucket(bucket)
+                                          .objects(prefix: stack)
+                                          .map(&:key)
```

Comment:
It already takes awhile to filter the logs so the more we can do in S3 the better.

Reply:
I don't think there's anything else we can do here with prefix?  the prefix search won't work past the v3 folder (in case they have v2 files, or we introduce v4 format), and then again the keyshasum might also change.

> /stack/v3/shared/keyshasum/.....

I don't think there's any way to search "deeper" during the S3 request.

Reply:
Ah, I forgot that the log version was in the string and yeah I'm pretty sure that S3 only supports prefix matching. I suppose we could do multiple prefix matches for the supported versions, etc but it's not a major concern at the moment.